src/Processors/Formats/Impl/ProtobufRowInputFormat.cpp

```diff
@@ -100,20 +139,162 @@ NamesAndTypesList ProtobufSchemaReader::readSchema()
 }
 
 /// proton: starts
+namespace
+{
+using ConfluentSchemaRegistry = ProtobufConfluentRowInputFormat::SchemaRegistryWithCache;
+#define SCHEMA_REGISTRY_CACHE_MAX_SIZE 1000
+/// Cache of Schema Registry URL + credentials -> SchemaRegistry
+LRUCache<std::string, ConfluentSchemaRegistry>  schema_registry_cache(SCHEMA_REGISTRY_CACHE_MAX_SIZE);
+
+std::shared_ptr<ConfluentSchemaRegistry> getConfluentSchemaRegistry(const String & base_url, const String & credentials)
+{
+    auto [schema_registry, loaded] = schema_registry_cache.getOrSet(
+        base_url + credentials,
+        [base_url, credentials]()
+        {
+            return std::make_shared<ConfluentSchemaRegistry>(base_url, credentials);
+        }
+    );
+    return schema_registry;
+}
+}
+
+class ProtobufConfluentRowInputFormat::SchemaRegistryWithCache: public google::protobuf::io::ErrorCollector
+{
+public:
+    SchemaRegistryWithCache(const String & base_url, const String & credentials)
+    : registry(base_url, credentials)
+    {
+    }
+
+    /// Overrides google::protobuf::io::ErrorCollector.
+    void AddError(int line, google::protobuf::io::ColumnNumber column, const std::string & message) override
+    {
+        throw Exception(ErrorCodes::INVALID_DATA, "Failed to parse schema, line={}, column={}, message={}", line, column, message);
+    }
+
+    const google::protobuf::Descriptor * getMessageType(uint32_t schema_id, const std::vector<Int64> & indexes)
+    {
+        assert(!indexes.empty());
+
+        const auto *fd = getSchema(schema_id);
+        auto mt_count = fd->message_type_count();
+        if (mt_count < indexes[0] + 1)
+            throw Exception(ErrorCodes::INVALID_DATA, "Invalid message index={} max_index={}", indexes[0], mt_count);
+
+        const auto *descriptor = fd->message_type(indexes[0]);
+
+        for (auto i : std::vector<Int64>(indexes.begin() + 1, indexes.end()))
```

Comment:
We may need put some comments here to explain the loop. It appears to me, we start from the initial index and follow the chain to get the last descriptor